CDAP-18010: Code refactoring to move parentClassloader creation into DefaultArtifactInspector

wyzhang opened the pull request:
This is part 2 of isolating artifact inspection.

Why:
Currently artifact inspection caller creates parent classloader and
then pass to inspector. This requires duplicating the logic of
parent classloader for local and remote execution.

Moving the parent classloader creation into DefaultArtifactInspector
avoids such duplication.

BEFORE
- artifact inspection
  CreateParentClassLoader()
    DefaultArtifactInspector(ArtifactId, parentClassLoader)
      // Inspect artifact

AFTER
- artifact inspection
  getParentArtifactDescriptor()
  DefaultArtifactInspector (ArtifactId, parent ArtifactDescriptor)
    CreateParentClassLoader(parent ArtifactDescriptor)
    // Inspect artifact

FUTURE
- local artifact inspection
    getParentArtifactDescriptor()
      DefaultArtifactInspector (ArtifactId, parent ArtifactDescriptor)
        CreateParentClassLoader(parent ArtifactDescriptor)
        // Inpsect artifact
- remote artifact inspection:
  getParentArtifactDescriptor()
    CallRemote(parent ArtifactDescriptor)
      DefaultArtifactInspector (ArtifactId, parent ArtifactDescriptor)
        CreateParentClassLoader(parent ArtifactDescriptor)
        // Inspect artifact

tivv commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/DefaultArtifactRepository.java:
```diff
@@ -531,16 +529,16 @@ private void addSystemArtifact(SystemArtifactInfo systemArtifactInfo) throws Exc
       LOG.warn("Could not add system artifact '{}' because it is invalid.", fileName, e);
     } catch (UnauthorizedException e) {
-      LOG.warn("Could not add system artifact '{}' because of an authorization error.", fileName, e);
+      LOG.warn("Coulde not add system artifact '{}' because of an authorization error.", fileName, e);
```
typo. Looks like accidental change

wyzhang replied:
Fixed.

tivv commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/ArtifactInspector.java:
```diff
@@ -34,16 +35,15 @@
    */
   ArtifactClassesWithMetadata inspectArtifact(Id.Artifact artifactId, File artifactFile,
-                                              @Nullable ClassLoader parentClassLoader,
+                                              @Nullable List<ArtifactDescriptor> parentDescriptors,
```
Please don't use a `Nullabble` List. Use empty list instead

wyzhang replied:
Done.

tivv commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/DefaultArtifactInspector.java:
```diff
@@ -106,43 +108,49 @@
   @Override
   public ArtifactClassesWithMetadata inspectArtifact(Id.Artifact artifactId, File artifactFile,
-                                                     @Nullable ClassLoader parentClassLoader,
+                                                     @Nullable List<ArtifactDescriptor> parentDescriptor,
```
Same `@Nullable` list over here

wyzhang replied:
Fixed.

wyzhang committed:
CDAP-18010: Code refactoring to move parentClassloader creation into DefaultArtifactInspector

This is part 2 of isolating artifact inspection.

Why:
Currently artifact inspection caller creates parent classloader and
then pass to inspector. This requires duplicating the logic of
parent classloader for local and remote execution.

Moving the parent classloader creation into DefaultArtifactInspector
avoids such duplication.

BEFORE
- artifact inspection
  CreateParentClassLoader()
     DefaultArtifactInspector(ArtifactId, parentClassLoader)
       // Inspect artifact

AFTER
- artifact inspection
  getParentArtifactDescriptor()
  DefaultArtifactInspector (ArtifactId, parent ArtifactDescriptor)
    CreateParentClassLoader(parent ArtifactDescriptor)
    // Inspect artifact

FUTURE
- local artifact inspection
    getParentArtifactDescriptor()
      DefaultArtifactInspector (ArtifactId, parent ArtifactDescriptor)
        CreateParentClassLoader(parent ArtifactDescriptor)
        // Inpsect artifact
- remote artifact inspection:
  getParentArtifactDescriptor()
    CallRemote(parent ArtifactDescriptor)
      DefaultArtifactInspector (ArtifactId, parent ArtifactDescriptor)
        CreateParentClassLoader(parent ArtifactDescriptor)
        // Inspect artifact
